relayer/app/leadstream.go

```diff
@@ -0,0 +1,281 @@
+package relayer
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/omni-network/omni/lib/cchain"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/log"
+	"github.com/omni-network/omni/lib/netconf"
+	"github.com/omni-network/omni/lib/umath"
+	"github.com/omni-network/omni/lib/xchain"
+
+	"github.com/tidwall/btree"
+)
+
+// errStop is returned when the leader should stop streaming.
+var errStop = errors.New("stop")
+
+// getLimit returns the limit of attestations to cache per chain version.
+func getLimit(network netconf.ID) int {
+	if network == netconf.Devnet {
+		return 20 // Tiny buffer in devnet
+	}
+
+	return 10_000 // 10k atts & 1KB per attestation & 10 chain versions ~= 100MB
+}
+
+// getBackoff returns the duration to backoff before querying the cache again.
+func getBackoff(network netconf.ID) time.Duration {
+	if network == netconf.Simnet {
+		return time.Millisecond // No backoff in tests
+	}
+
+	return time.Second // Default 1 second blocks otherwise
+}
+
+// leadChaosTimeout returns a function that returns true if the leader chaos timeout has been reached.
+// This ensures we rotate leaders after a certain time (and test leader rotation).
+func leadChaosTimeout(network netconf.ID) func() bool {
+	t0 := time.Now()
+	return func() bool {
+		duration := time.Hour // Default 1 hour timeout
+		if network == netconf.Devnet {
+			duration = time.Second * 10
+		} else if network == netconf.Staging {
+			duration = time.Minute
+		}
+
+		return time.Since(t0) > duration
+	}
+}
+
+// leader tracks a worker actively streaming attestations and adding them to the cache.
+// It "locks" a range of offsets as leader of that range.
+type leader struct {
+	mu     sync.RWMutex
+	from   uint64
+	latest uint64
```

Comment:
It feels like `latest` is not really needed because it's only used to _guess_ if the range was covered by somebody else or not (this does not directly prove that the range is consistenty filled). So in case of some invariant violation (for whatever reason) we can get a leader with inconsistent range coverage that will prevent others from starting on its range.

Instead we could not rely on ranges at all and simply start a new leader on _any_ height that is a cache miss — it will ultimately bump into somebody's range (or timeout).

Reply:
nope, this is actually critical, since, latest is increased as soon as previous is fetched, this ensure that this leader remains the leader and others don't also start streaming from next as soon as it isn't in the cache.